src/Databases/DatabasesCommon.cpp

```diff
@@ -125,16 +126,43 @@ void DatabaseWithOwnTablesBase::shutdown()
         tables_snapshot = tables;
     }
 
-    for (const auto & kv : tables_snapshot)
+    auto shutdown_table = [&](const StoragePtr & table)
     {
-        auto table_id = kv.second->getStorageID();
-        kv.second->shutdown();
+        auto table_id = table->getStorageID();
+        table->shutdown();
         if (table_id.hasUUID())
         {
             assert(getDatabaseName() == DatabaseCatalog::TEMPORARY_DATABASE || getUUID() != UUIDHelpers::Nil);
             DatabaseCatalog::instance().removeUUIDMapping(table_id.uuid);
         }
+    };
+
+    /// Destroy Buffer tables first, to flush the data to the underlying storage.
+    std::vector<StoragePtr> buffer_storages;
+    std::vector<StoragePtr> other_storages;
+
+    for (const auto & [_, storage] : tables_snapshot)
+    {
+        if (dynamic_cast<StorageBuffer *>(storage.get()))
+            buffer_storages.push_back(storage);
+        else
+            other_storages.push_back(storage);
+    }
+
+    if (!buffer_storages.empty())
+    {
+        LOG_INFO(log, "Found {} Buffer tables. Destroying them before others.",
+            buffer_storages.size());
+
+        Stopwatch watch;
+        for (const auto & storage : buffer_storages)
+            shutdown_table(storage);
+
+        UInt64 ms = watch.elapsedMilliseconds();
+        LOG_INFO(log, "Buffer tables had been destroyed (took {} ms).", ms);
```

Comment:
I am not sure that we need such log, if we does not log other tables shutdown.

Reply:
The reason for logging here is that flushing buffer to the underlying table may take significant amount of time

Reply:
Agree, this log message is superfluous, since Buffer will have message in flush anyway.